### PR TITLE
[feat] Adding Visual Attention

### LINKED ISSUE #319: Support Visual Attention Network
# 🚀 Feature
Add support for https://arxiv.org/abs/2202.09741

## Motivation
Superb results supported, pretty different structure when compared to some other models so would be nice to check whether the abstractions work out

## Pitch
Support it in parts, as a follow up of the Poolformers integration

## Alternatives
Not doing it

## Additional context
Official implementation: https://github.com/Visual-Attention-Network

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Fixes #319. Note that to reproduce the paper you need the Conv2DFeedforward introduced here #321, and a metaformer-like structure

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?
- [x] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
cc @mannatsingh if you're interested in these things

**Comment by blefaudeux** [comment]
training a 8M metaformer-derived model with the visual attention from this PR: (for the record resnet18 is 12M params and goes to 93%)

File: examples/cifarMetaformer.py [hunk 1]
```diff
@@ -121,8 +121,8 @@ def forward(self, x):
 
     # Adjust batch depending on the available memory on your machine.
     # You can also use reversible layers to save memory
-    REF_BATCH = 512
-    BATCH = 512  # lower if not enough GPU memory
+    REF_BATCH = 768
```
Comment (blefaudeux) [inline examples/cifarMetaformer.py:124]: looks like a classic default for Cifar10

File: examples/cifarMetaformer.py [hunk 2]
```diff
@@ -31,6 +31,7 @@ def __init__(
         num_classes=10,
         dim=384,
         attention="scaled_dot_product",
+        feedforward="MLP",
```
Comment (blefaudeux) [inline examples/cifarMetaformer.py:34]: I'm not sure about the defaults here, how to show that you can use these to repro "Visual Attention" for instance ? Should we show different presets  ?

File: xformers/components/attention/pooling.py [hunk 3]
```diff
@@ -62,6 +62,10 @@ def __init__(
         # This operator does not really handle q,k,v
         self.requires_same_k_q_dimensions = True
 
+        # This attention requires the 2d structure out of the context,
+        # implictly assumed to be a squared length
+        self.requires_squared_context = True
```
Comment (blefaudeux) [inline xformers/components/attention/pooling.py:67]: this was already true before, but not formalized like this, I think it's cleaner ? "pooling" (PoolingFormer) and "visual" both recover the 2d structure of and assume a squared context length for that

File: xformers/components/attention/visual.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            False  # This mechanism does not require that the MHA projects inputs
+        )
+
+    def forward(self, q: torch.Tensor, *_, **__):
+        # Expose the 2D token structure
+        B, HW, C = q.shape
+        H = int(math.sqrt(HW))
+        assert H * H == HW
+
+        x = q.transpose(-2, -1).reshape(B, C, H, H)
```
Comment (blefaudeux) [inline xformers/components/attention/visual.py:88]: I've not benchmarked that, but maybe that it's beneficial to .contiguous() here, depending on the Conv2D kernels

**Review by dianaml0 — APPROVED** [review]

**Pull request merged**